### ISSUE TITLE
Count api suggestion/potential improvement?

### ISSUE THREAD

**Issue by Lux-Delux** [jacobwb/hashover-next#104]
Currently the count api can't be deferred. Not a BIG deal but it still manages to drop pagespeed scores by 2 points (yeah yeah I know it's nit picking but still...). 

It's obvious we can't use document.write after page load so...

Perhaps make it hook onto an element like the comments, with e.g. id="hashover-api-count" or something. So the js only outputs the text but you need a HTML href element on which we can then use

`document.getElementById('hashover-api-count').innerHTML = count_link;`

In effect the api doesn't slow down page load and makes pagespeed happy. Thoughts?

**Comment by jacobwb** [comment]
The `hashover.js` file should be able to be deferred. While the `hashover/api/count_link.php` file that `hashover.js` loads is not. Such is the design of JavaScript.

You should be able to defer the count API by linking to it directly, though.

```html
<script id="hashover-script-count" type="text/javascript" defer src="/hashover/api/count_link.php?url=http://example.com/test.html&hashover-script=count"></script>
```

I thought deferring JavaScript is not normally preferred, and that asynchronous scripts perform better. Since deferred JavaScript files are downloaded alongside everything else, but executed once the page is done loading, whereas asynchronous scripts download alongside everything else, without blocking HTML rendering, and execute as soon as they finish downloading.

I don't quite understand what you're trying to do though. Are you trying to defer multiple count API script tags, that are, like, on the homepage? Or are you trying to place a count link someplace else on the same page as the comments? There is a method to do the latter more efficiently.

**Comment by Lux-Delux** [comment]
I guess I used the term 'defer' too freely.

What I basically mean is: prevent the count.api from slowing down page load. Be that with defer or async.

I'm using the count in my sidebar.

**Comment by jacobwb** [comment]
The normal HashOver comments script will automatically place the number of comments into an HTML element with the ID "hashover-cmtcount".

For example, I do it like this:

```html
<span style="float: right;">
    <a href="#comments"><span id="hashover-cmtcount"></span> Comments</a> &#8628;
</span>
```

So something similar to the count link API would be:

```html
<a href="#comments"><span id="hashover-cmtcount">Post</span> Comments</a>
```

It's somewhat more limited than the PHP count link API in that you can't easily display "Comment" when there is only one comment and "Comments" when there are more or less than one comment.

But it does allow for things like this:

![screenshot from 2015-04-02 20 15 15](https://cloud.githubusercontent.com/assets/4677622/6977629/6657c226-d976-11e4-9529-d59964f9d83c.png)

**Comment by Lux-Delux** [comment]
Ah! Perfect, exactly what I needed.

BTW you can achieve the same as the above image with both methods using css pseudo selectors.

**Comment by Lux-Delux** [comment]
I'm experimenting adding comments to, for example, the homepage.

Let's say it lists 10 posts and I'd like to have a link on each with the comment count.

What would currently be the best solution for this performance wise? So that each post gets it's own count and proper link?

**Comment by jacobwb** [comment]
The only way currently is to place multiple JavaScript tags pointing to the count link API.

I'll try to implement something better soon.

It will probably look like this:

```html
<article title>
<a href="/path/to/page" class="hashover-count-link">Post Comment</a>
<article summary>

<article title>
<a href="/path/to/page" class="hashover-count-link">Post Comment</a>
<article summary>

<article title>
<a href="/path/to/page" class="hashover-count-link">Post Comment</a>
<article summary>
...
<script type="text/javascript" src="/hashover.js?api=count_links"></script>
```

The JavaScript will pull each anchor tag with the "hashover-count-link" class and hand their URLs to the PHP, which will fetch a comment count for each page and output JavaScript to place the comment counts into their respective elements.

That should be much more efficient, in both the fact that multiple requests to the same file wouldn't be made, and the PHP wouldn't be executing multiple times either.
